app/drivers/distributed_manager.go

```diff
@@ -64,6 +183,192 @@ func (d *DistributedManager) IsDistributed() bool {
 	return true
 }
 
+// provisionFromPool overrides the Manager's provisionFromPool method to use FindAndClaim for distributed coordination
+func (d *DistributedManager) provisionFromPool(
+	ctx context.Context,
+	pool *poolEntry,
+	tlsServerName, ownerID, resourceClass string,
+	vmImageConfig *spec.VMImageConfig,
+	agentConfig *types.GitspaceAgentConfig,
+	storageConfig *types.StorageConfig,
+	zone, machineType string,
+	shouldUseGoogleDNS bool,
+	timeout int64,
+	poolName string,
+) (*types.Instance, bool, error) {
+	// For distributed manager, first try to claim an existing free instance
+	allowedStates := []types.InstanceState{types.StateCreated}
+
+	// Try to find and claim a free instance atomically
+	inst, err := d.instanceStore.FindAndClaim(ctx, &types.QueryParams{PoolName: poolName}, types.StateInUse, allowedStates)
+	if err != nil && err != sql.ErrNoRows {
+		return nil, false, fmt.Errorf("provision: failed to find and claim instance in %q pool: %w", poolName, err)
+	}
+
+	// If we successfully claimed an instance, update it and return
+	if inst != nil {
+		inst.OwnerID = ownerID
+		if inst.IsHibernated {
```

Comment:
Shouldnt we have types.StateHibernated in allowedStates above if we think we will hit this condition? We also use setupInstanceWithHibernate so we hibernate new instances set up in the pool right?